File: sdk/src/main/java/com/meniga/sdk/webservices/requests/GetFeed.java
package com.meniga.sdk.webservices.requests;

import org.joda.time.DateTime;
import org.joda.time.format.DateTimeFormatter;
import org.joda.time.format.ISODateTimeFormat;

import java.util.HashMap;
import java.util.Map;

/**
 * Copyright 2017 Meniga Iceland Inc.
 */
public class GetFeed extends QueryRequestObject {

	public DateTime dateFrom;
	public DateTime dateTo;
	public int skip = 0;
	public int take = -1;

	@Override
	public long getValueHash() {
		int result = dateFrom != null ? dateFrom.hashCode() : 0;
		result = 31 * result + (dateTo != null ? dateTo.hashCode() : 0);
		result = 31 * result + skip;
		result = 31 * result + take;
		return result;
	}

	@Override
	public Map<String, String> toQueryMap() {
		Map<String, String> query = new HashMap<>();

		DateTimeFormatter formatter = ISODateTimeFormat.dateTime();

		query.put("dateFrom", formatter.print(dateFrom));
		query.put("dateTo", formatter.print(dateTo));

		if (take > 0) {
			query.put("skip", Integer.toString(skip));
			query.put("take", Integer.toString(take));
		}

		return query;
	}
}

File: sdk/src/main/java/com/meniga/sdk/webservices/requests/GetFeed.kt
/*
 * Copyright 2017-2021 Meniga Iceland Inc.
 */
package com.meniga.sdk.webservices.requests

import com.meniga.sdk.helpers.toSkipAndTake
import com.meniga.sdk.models.feed.FeedFilter
import org.joda.time.format.ISODateTimeFormat

data class GetFeed(
        @JvmField var feedFilter: FeedFilter
) : QueryRequestObject() {

    override fun toQueryMap(): MutableMap<String, String> =
            mutableMapOf<String, String>().also { queryMap ->
                val formatter = ISODateTimeFormat.dateTime()
                feedFilter.from.let { queryMap["dateFrom"] = formatter.print(it) }
                feedFilter.to.let { queryMap["dateTo"] = formatter.print(it) }
                feedFilter.type?.let { queryMap["type"] = it.joinToString(",") }
                feedFilter.eventTypeIdentifiers?.let { queryMap["eventTypeIdentifiers"] = it.joinToString(",") }
                val (skip, take) = toSkipAndTakeIfProvided(feedFilter)
                skip?.let { queryMap["skip"] = it.toString() }
                take?.let { queryMap["take"] = it.toString() }
            }

    override fun getValueHash() = hashCode().toLong()

    private fun toSkipAndTakeIfProvided(feedFilter: FeedFilter) =
            if (feedFilter.page != null && feedFilter.itemsPerPage != null) {
                toSkipAndTake(feedFilter.page, feedFilter.itemsPerPage)
            } else {
                Pair(null, null)
            }
}
